makeflow/src/mountfile.c

```diff
@@ -0,0 +1,660 @@
+	if(!link) {
+		LDEBUG("http_query(%s, \"HEAD\", ...) failed!\n", url);
+		fprintf(stderr, "http_query(%s, \"HEAD\", ...) failed!\n", url);
+		return -1;
+	}
+	return 0;
+}
+
+/* mount_install_local copies source to target.
+ * @param source: a local file path which must exist already.
+ * @param target: a local file path which must not exist.
+ * @param cache_path: a file path in the cache dir.
+ * @param s_type: the file type of source.
+ * return 0 on success; return -1 on failure.
+ */
+int mount_install_local(const char *source, const char *target, const char *cache_path, file_type s_type) {
+	switch(s_type) {
+	case FILE_TYPE_REG:
+		if(copy_file_to_file(source, cache_path) < 0) {
+			LDEBUG("copy_file_to_file from %s to %s failed.\n", source, cache_path);
+			return -1;
+		}
+		break;
+	case FILE_TYPE_LNK:
+		if(copy_symlink(source, cache_path)) {
+			LDEBUG("copy_symlink from %s to %s failed.\n", source, cache_path);
+			return -1;
+		}
+	case FILE_TYPE_DIR:
+		if(copy_dir(source, cache_path)) {
+			LDEBUG("copy_dir from %s to %s failed.\n", source, cache_path);
+			return -1;
+		}
+		break;
+	default:
+		break;
+	}
+
+	return 0;
+}
+
+/* mount_check checks the validity of source and target.
+ * It also sets the file type when source is a local path.
+ * @param source: the source location of a dependency, which can be a local file path or http URL.
+ * @param target: a local file path which must not exist.
+ * @param s_type: the file type of source
+ * return 0 on success; return -1 on failure.
+ */
+int mount_check(const char *source, const char *target, file_type *s_type) {
+	if(!source || !*source) {
+		LDEBUG("the source (%s) can not be empty!\n", source);
+		fprintf(stderr, "the source (%s) can not be empty!\n", source);
+		return -1;
+	}
+
+	if(!target || !*target) {
+		LDEBUG("the target (%s) can not be empty!\n", target);
+		fprintf(stderr, "the target (%s) can not be empty!\n", target);
+		return -1;
+	}
+
+	/* Check whether the target is an absolute path. */
+	if(target[0] == '/') {
+		LDEBUG("the target (%s) should not be an absolute path!\n", target);
+		fprintf(stderr, "the target (%s) should not be an absolute path!\n", target);
+		return -1;
+	}
+
+	/* check whether target includes .. */
+	if(path_has_doubledots(target)) {
+		LDEBUG("the target (%s) include ..!\n", target);
+		fprintf(stderr, "the target (%s) include ..!\n", target);
+		return -1;
+	}
+
+	/* check whether target includes any symlink link, this check prevent the makeflow breaks out the CWD. */
+	if(path_has_symlink(target)) {
+		LDEBUG("the target (%s) should not include any symbolic link!\n", target);
+		fprintf(stderr, "the target (%s) should not include any symbolic link!\n", target);
+		return -1;
+	}
+
+	/* Check whether target already exists. */
+	if(!access(target, F_OK)) {
+		LDEBUG("the target (%s) already exists!\n", target);
+		fprintf(stderr, "the target (%s) already exists!\n", target);
+		return -1;
```

Comment:
This if statement prevents the user from using the same commandline, even if the mountfile is the same. I thought we only prevent them from moving forward if the mountfiles were different and changed the information?